src/metabase/api/dashboard.clj

```diff
@@ -836,6 +842,54 @@
     :user-id     api/*current-user-id*
     :revision-id revision_id}))
 
+(defn- dashboard-metadata
+  [dashboard]
+  (let [dashcards (:dashcards dashboard)
+        cards (for [{:keys [card series]} dashcards
+                    :let [all (conj series card)]
+                    card all
+                    :when card]
+                card)
+        database-ids (map :database_id cards)
+        db->mp (into {} (map (juxt identity lib.metadata.jvm/application-database-metadata-provider)
+                             database-ids))
+        dependents (group-by :type (set (mapcat (fn [card]
+                                                  (let [database-id (:database_id card)
+                                                        mp (db->mp database-id)
+                                                        query (lib/query mp (:dataset_query card))]
+                                                    (lib/dependent-metadata query (:id card) (:type card)))) cards)))]
+    {:tables (into []
+                   (keep (fn [{card-or-table-id :id}]
+                           (try
+                             (if-let [card-id (lib.util/legacy-string-table-id->card-id card-or-table-id)]
+                               (api.table/fetch-card-query-metadata card-id)
+                               (api.table/fetch-table-query-metadata card-or-table-id {}))
+                             (catch Exception e
+                               (log/warnf "Error in dashboard metadata %s %s: %s" :table card-or-table-id (ex-message e))))))
+                   (:table dependents))
+     :databases (into []
+                      (keep (fn [{database-id :id}]
+                              (try
+                                (api.database/get-database database-id {})
+                                (catch Exception e
+                                  (log/warnf "Error in dashboard metadata %s %s: %s" :database database-id (ex-message e))))))
+                      (:database dependents))
+     :fields (into []
+                   (keep (fn [{field-id :id}]
+                           (try
+                             (api.field/get-field field-id {})
+                             (catch Exception e
+                               (log/warnf "Error in dashboard metadata %s %s: %s" :field field-id (ex-message e))))))
+                   (:field dependents))}))
```

Comment:
I suppose we could reduce the number of DB requests too. (Probably in a followup and depending on how effective this optimization turns out to be.)

Reply:
The amount of data and duplication here is kinda unreal. Reducing the number of db requests and deduplicating the data across all the structures would be great but probably a large effort to tease the pieces apart.

I think, ideally we change it to go from dependents to a fully populated metadata-provider, dump it over the wire and update the FE so it can build a metadata provider with dump.